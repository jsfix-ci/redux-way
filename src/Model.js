export default class Model {
	constructor ({state, constants, modelName} = {}) {
		this.state = state;
		this.modelName = modelName;

		Model.self = () => (this);
	}

	update (state) {
		if (this.state instanceof Object && !Array.isArray(this.state)) {
			this.state = { ...this.state, ...state }
		} else {
			this.state = state;
		}
	}

	run = function* (){}
}