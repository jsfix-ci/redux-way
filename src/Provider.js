import React from 'react'
import PropTypes from 'prop-types'

export default class Provider extends React.Component {
	static propTypes = {
		store: PropTypes.object.isRequired
	}

	componentWillMount() {
		
	}
}