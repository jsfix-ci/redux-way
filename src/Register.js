export default class Register {
	constructor() {
		this.models = [];
		this.saga = false;
	}

	register(...models) {
		models.forEach((model) => {
			this.models.push(new model({
				state: model.state,
				constants: model.constants,
				modelName: model.modelName
			}))
		})
	}

	sagaMiddleware(saga) {
		this.saga = saga;
		this.models.forEach((model) => {
			if (model.run) {
				this.saga.run(model.run);
			}
		});
	}

	getState() {
		let state = {};
		for(let index in this.models) {
			let model = this.models[index];
			state[model.modelName] = model.state;
		}
		return state;
	}
}