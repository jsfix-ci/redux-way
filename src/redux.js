import React from 'react'
import PropTypes from 'prop-types';
import isEmpty from 'lodash/isEmpty'
import extend from 'lodash/extend'

export const createReducer = (register) => 
	(state, action) => {
		register.models.forEach((model) => {
			for (let type in model.reducer) {
				if (type === action.type && typeof model.reducer[type] === 'function') {
					model.reducer[type](model.state, action, model);
				}
			}
		})
		return register.getState()
	}

export const connect = (state, actions = {}) => {
	return (Child) => {
		class WrappedComponent extends React.Component {
			static contextTypes = {
				store: PropTypes.object
			}

			constructor(props, context) {
				super(props, context);
				this.state = {};

				let setState = typeof state === 'function' ? state(this.context.store.getState()) : {};
				for (let name in actions) {
					setState[name] = (e, ...params) => {
						this.context.store.dispatch(actions[name](...params))
					};
				}
				this.state = setState
			}

			componentWillMount() {
				this.subscribe = this.context.store.subscribe(() => {
					let setState = typeof state === 'function' ? state(this.context.store.getState()) : {};
					let newState = {};
					for (let key in setState) {
						if (this.state.hasOwnProperty(key) && JSON.stringify(this.state[key]) !== JSON.stringify(setState[key])) {
							newState[key] = setState[key]
						}
					}
					if (!isEmpty(newState) && this.mount) {
						this.setState(newState);
					}
				})
			}

			componentWillUnmount() {
				this.mount = false;
				if (this.subscribe && this.subscribe.unsubscribe && typeof this.subscribe.unsubscribe === 'function') {
					this.subscribe.unsubscribe()
				}
			}

			componentDidMount() {
				this.mount = true;
			}

			render() {
				return React.createElement(Child, {...this.state, ...this.props})
			}
		}

		return WrappedComponent;
	}
}