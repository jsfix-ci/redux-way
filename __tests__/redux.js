import ReactTestRenderer from 'react-test-renderer';
import React from 'react'
import PropTypes from 'prop-types'
import {createStore} from 'redux'

import {createReducer, connect} from '../src/redux';

describe('redux', () => {
	describe('createRedux', () => {
		it('must be a function', () => {
			expect(typeof createReducer).toBe('function')
		});

		it('must return function', () => {
			expect(typeof createReducer()).toBe('function')
		});

		it('must return function', () => {
			const reducer = {
				getState: jest.fn(() => (1)),
				models: [{
					reducer:{'test': jest.fn(() => (1))},
					state: 5
				}]
			};
			const state = {};
			const action = {
				type: "test"
			};
			expect(createReducer(reducer)(state, action)).toBe(1)
			expect(reducer.models[0].reducer.test).toHaveBeenCalled()
			expect(reducer.getState).toHaveBeenCalled()
		});
	})
	
	describe('connect', () => {
		it('must be a function', () => {
			expect(typeof connect).toBe('function')
		});

		it('must return function', () => {
			expect(typeof connect()).toBe('function')
		});

		it('must return class extends by react', () => {
			const action = jest.fn();
			const reducer = jest.fn()
			const store = createStore(state=>state)

			class Provider extends React.Component {
				static childContextTypes = {
					store: PropTypes.object.isRequired
				}
			 
				static propTypes = {
					store: PropTypes.object.isRequired
				}
			 
				getChildContext() {
					return {store: this.props.store};
				}
			 
				render() {
					return this.props.children
				}
			}

			class Test extends React.Component{
				render(){return null}
			}
			const Component = connect(() => ({}), {action: action})(Test)

			const renderer = ReactTestRenderer.create(
				<Provider store={store}>
					<Component />
				</Provider>
			)
			
			const json = renderer.toJSON();
	
			store.dispatch({type:''});
			expect(json).toMatchSnapshot();
		});
	})
})