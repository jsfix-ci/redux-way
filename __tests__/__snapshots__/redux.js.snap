// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux connect must return class extends by react 1`] = `null`;
